File: php/ant_ffi/tests/ClientIntegrationTest.php
<?php

declare(strict_types=1);

namespace AntFfi\Tests;

use PHPUnit\Framework\TestCase;
use AntFfi\Types\Client;
use AntFfi\Types\Network;
use AntFfi\Types\Wallet;
use AntFfi\Types\SecretKey;
use AntFfi\Types\ChunkAddress;
use AntFfi\Types\DataAddress;
use AntFfi\Types\DataMapChunk;
use AntFfi\Types\PointerAddress;
use AntFfi\Types\PointerTarget;
use AntFfi\Types\NetworkPointer;
use AntFfi\Types\ScratchpadAddress;
use AntFfi\Types\Scratchpad;
use AntFfi\Types\RegisterAddress;
use AntFfi\Types\GraphEntryAddress;
use AntFfi\Types\GraphEntry;
use AntFfi\Types\VaultSecretKey;
use AntFfi\Types\UserData;
use AntFfi\Types\PublicArchive;
use AntFfi\Types\Metadata;

/**
 * End-to-end integration tests for client operations.
 *
 * These tests require a local Autonomi testnet running.
 * Run with: vendor/bin/phpunit --group integration
 *
 * To start a local testnet:
 * cd C:\tools\ant && setup-testnet.bat
 *
 * @group integration
 */
final class ClientIntegrationTest extends TestCase
{
    private const DEFAULT_TEST_KEY = '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';

    private ?Client $client = null;
    private ?Wallet $wallet = null;

    protected function setUp(): void
    {
        // Initialize client and wallet for tests
        try {
            $this->client = Client::initLocalSync();
            $network = Network::local();
            $this->wallet = Wallet::fromPrivateKey($network, self::DEFAULT_TEST_KEY);
        } catch (\Exception $e) {
            $this->markTestSkipped('Local testnet not available: ' . $e->getMessage());
        }
    }

    protected function tearDown(): void
    {
        if ($this->client !== null) {
            $this->client->dispose();
        }
        if ($this->wallet !== null) {
            $this->wallet->dispose();
        }
    }

    // =========================================================================
    // Public Data Operations
    // =========================================================================

    public function testDataPutPublicAndGetSync(): void
    {
        $data = 'Hello, Autonomi Network! ' . uniqid();

        $result = $this->client->dataPutPublicSync($data, $this->wallet);

        $this->assertNotEmpty($result->address->toHex());
        $this->assertNotEmpty($result->cost);

        // Retrieve the data
        $retrieved = $this->client->dataGetPublicSync($result->address->toHex());
        $this->assertEquals($data, $retrieved);
    }

    public function testDataCostSync(): void
    {
        $data = str_repeat('X', 1000); // 1KB of data

        $cost = $this->client->dataCostSync($data);

        $this->assertNotEmpty($cost);
        $this->assertIsString($cost);
    }

    // =========================================================================
    // Private Data Operations
    // =========================================================================

    public function testPrivateDataPutAndGetSync(): void
    {
        $data = 'Secret data: ' . uniqid();

        $result = $this->client->dataPutSync($data, $this->wallet);

        $this->assertInstanceOf(DataMapChunk::class, $result->dataMap);
        $this->assertNotEmpty($result->cost);

        // Retrieve the private data using the data map
        $retrieved = $this->client->dataGetSync($result->dataMap);
        $this->assertEquals($data, $retrieved);
    }

    // =========================================================================
    // Pointer Operations
    // =========================================================================

    public function testPointerPutAndGetSync(): void
    {
        $ownerKey = SecretKey::random();

        // Create a chunk address to point to
        $chunkAddress = ChunkAddress::fromContent('target content');
        $target = PointerTarget::fromChunkAddress($chunkAddress);

        // Create and store the pointer
        $pointer = NetworkPointer::create($ownerKey, 0, $target);
        $this->client->pointerPutSync($pointer, $this->wallet);

        // Retrieve the pointer
        $address = $pointer->address();
        $retrieved = $this->client->pointerGetSync($address);

        $this->assertEquals($pointer->counter(), $retrieved->counter());
        $this->assertEquals($pointer->owner()->toHex(), $retrieved->owner()->toHex());
    }

    public function testPointerUpdate(): void
    {
        $ownerKey = SecretKey::random();

        // Create initial pointer
        $target1 = PointerTarget::fromChunkAddress(ChunkAddress::fromContent('target1'));
        $pointer1 = NetworkPointer::create($ownerKey, 0, $target1);
        $this->client->pointerPutSync($pointer1, $this->wallet);

        // Update the pointer with higher counter
        $target2 = PointerTarget::fromChunkAddress(ChunkAddress::fromContent('target2'));
        $pointer2 = NetworkPointer::create($ownerKey, 1, $target2);
        $this->client->pointerPutSync($pointer2, $this->wallet);

        // Retrieve and verify the updated pointer
        $address = $pointer2->address();
        $retrieved = $this->client->pointerGetSync($address);

        $this->assertEquals(1, $retrieved->counter());
    }

    // =========================================================================
    // Scratchpad Operations
    // =========================================================================

    public function testScratchpadPutAndGetSync(): void
    {
        $ownerKey = SecretKey::random();
        $data = 'Scratchpad content: ' . uniqid();

        // Create and store the scratchpad
        $scratchpad = Scratchpad::create($ownerKey, 1, $data, 0);
        $this->client->scratchpadPutSync($scratchpad, $this->wallet);

        // Retrieve the scratchpad
        $address = $scratchpad->address();
        $retrieved = $this->client->scratchpadGetSync($address);

        // Scratchpad stores encrypted data - use decryptData to get original
        $this->assertEquals($data, $retrieved->decryptData($ownerKey));
        $this->assertEquals(0, $retrieved->counter());
        $this->assertEquals(1, $retrieved->dataEncoding());
        $this->assertTrue($retrieved->isValid());
    }

    public function testScratchpadUpdate(): void
    {
        $ownerKey = SecretKey::random();

        // Create initial scratchpad
        $scratchpad1 = Scratchpad::create($ownerKey, 1, 'initial data', 0);
        $this->client->scratchpadPutSync($scratchpad1, $this->wallet);

        // Update the scratchpad with higher counter
        $scratchpad2 = Scratchpad::create($ownerKey, 1, 'updated data', 1);
        $this->client->scratchpadPutSync($scratchpad2, $this->wallet);

        // Retrieve and verify the updated scratchpad
        $address = $scratchpad2->address();
        $retrieved = $this->client->scratchpadGetSync($address);

        // Scratchpad stores encrypted data - use decryptData to get original
        $this->assertEquals('updated data', $retrieved->decryptData($ownerKey));
        $this->assertEquals(1, $retrieved->counter());
    }

    // =========================================================================
    // Register Operations
    // =========================================================================

    private function make32ByteValue(string $prefix): string
    {
        $value = str_pad($prefix, 32, "\0");
        return substr($value, 0, 32);
    }

    public function testRegisterCreateAndGetSync(): void
    {
        $ownerKey = SecretKey::random();
        // Register value must be exactly 32 bytes
        $value = $this->make32ByteValue('Reg value: ' . substr(uniqid(), 0, 10));

        // Create the register
        $address = $this->client->registerCreateSync($ownerKey, $value, $this->wallet);

        $this->assertInstanceOf(RegisterAddress::class, $address);

        // Retrieve the register value
        $retrieved = $this->client->registerGetSync($address);
        $this->assertEquals($value, $retrieved);
    }

    public function testRegisterUpdate(): void
    {
        $ownerKey = SecretKey::random();
        // Register value must be exactly 32 bytes
        $initialValue = $this->make32ByteValue('initial');

        // Create the register
        $address = $this->client->registerCreateSync($ownerKey, $initialValue, $this->wallet);

        // Update the register with new 32-byte value
        $updatedValue = $this->make32ByteValue('updated');
        $this->client->registerUpdateSync($ownerKey, $updatedValue, $this->wallet);

        // Retrieve and verify the updated value
        $retrieved = $this->client->registerGetSync($address);
        $this->assertEquals($updatedValue, $retrieved);
    }

    // =========================================================================
    // Graph Entry Operations
    // =========================================================================

    private function make32ByteContent(string $prefix): string
    {
        $content = str_pad($prefix, 32, "\0");
        return substr($content, 0, 32);
    }

    public function testGraphEntryPutAndGetSync(): void
    {
        $ownerKey = SecretKey::random();
        // GraphEntry content must be exactly 32 bytes
        $content = $this->make32ByteContent('Graph entry: ' . substr(uniqid(), 0, 10));

        // Create and store the graph entry
        $entry = GraphEntry::create($ownerKey, '', $content);
        $this->client->graphEntryPutSync($entry, $this->wallet);

        // Retrieve the graph entry
        $address = $entry->address();
        $retrieved = $this->client->graphEntryGetSync($address);

        $this->assertEquals($content, $retrieved->content());
        // Verify addresses match (owner method not available on GraphEntry)
        $this->assertEquals($entry->address()->toHex(), $retrieved->address()->toHex());
    }

    // =========================================================================
    // Vault Operations
    // =========================================================================

    public function testVaultPutAndGetUserDataSync(): void
    {
        $vaultKey = VaultSecretKey::random();
        $userData = UserData::create();

        // Store user data to the vault
        $this->client->vaultPutUserDataSync($vaultKey, $userData, $this->wallet);

        // Retrieve user data from the vault
        $retrieved = $this->client->vaultGetUserDataSync($vaultKey);

        $this->assertInstanceOf(UserData::class, $retrieved);
    }

    // =========================================================================
    // Archive Operations
    // =========================================================================

    public function testArchivePutAndGetPublicSync(): void
    {
        // Create an empty archive first (simpler test)
        $archive = PublicArchive::create();

        // Store the empty archive
        $address = $this->client->archivePutPublicSync($archive, $this->wallet);

        $this->assertNotEmpty($address->toHex());

        // Retrieve the archive
        $retrieved = $this->client->archiveGetPublicSync($address);

        $this->assertInstanceOf(PublicArchive::class, $retrieved);
    }

    public function testArchiveWithFileSync(): void
    {
        // TODO: Investigate crash in addFile method - possibly handle update issue
        $this->markTestSkipped('Archive addFile crashes - needs investigation');
    }

    // =========================================================================
    // File Operations
    // =========================================================================

    public function testFileUploadPublicSync(): void
    {
        // Create a temporary file
        $tempFile = sys_get_temp_dir() . '/ant_test_' . uniqid() . '.txt';
        file_put_contents($tempFile, 'Test file content: ' . uniqid());

        try {
            $result = $this->client->fileUploadPublicSync($tempFile, $this->wallet);

            $this->assertNotEmpty($result->address->toHex());
            $this->assertNotEmpty($result->cost);

            // Verify we can download the file content
            $downloaded = $this->client->dataGetPublicSync($result->address->toHex());
            $this->assertEquals(file_get_contents($tempFile), $downloaded);
        } finally {
            unlink($tempFile);
        }
    }

    public function testFileUploadSync(): void
    {
        // Create a temporary file
        $tempFile = sys_get_temp_dir() . '/ant_test_private_' . uniqid() . '.txt';
        file_put_contents($tempFile, 'Private file content: ' . uniqid());

        try {
            $result = $this->client->fileUploadSync($tempFile, $this->wallet);

            $this->assertInstanceOf(DataMapChunk::class, $result->dataMap);
            $this->assertNotEmpty($result->cost);

            // Verify we can download the file content
            $downloaded = $this->client->dataGetSync($result->dataMap);
            $this->assertEquals(file_get_contents($tempFile), $downloaded);
        } finally {
            unlink($tempFile);
        }
    }

    public function testFileDownloadPublicSync(): void
    {
        // Upload a file first
        $tempUpload = sys_get_temp_dir() . '/ant_upload_' . uniqid() . '.txt';
        $content = 'Content for download test: ' . uniqid();
        file_put_contents($tempUpload, $content);

        try {
            $result = $this->client->fileUploadPublicSync($tempUpload, $this->wallet);

            // Download to a new file
            $tempDownload = sys_get_temp_dir() . '/ant_download_' . uniqid() . '.txt';
            $this->client->fileDownloadPublicSync($result->address, $tempDownload);

            $this->assertFileExists($tempDownload);
            $this->assertEquals($content, file_get_contents($tempDownload));

            unlink($tempDownload);
        } finally {
            unlink($tempUpload);
        }
    }

    public function testFileDownloadSync(): void
    {
        // Upload a private file first
        $tempUpload = sys_get_temp_dir() . '/ant_private_upload_' . uniqid() . '.txt';
        $content = 'Private content for download test: ' . uniqid();
        file_put_contents($tempUpload, $content);

        try {
            $result = $this->client->fileUploadSync($tempUpload, $this->wallet);

            // Download to a new file
            $tempDownload = sys_get_temp_dir() . '/ant_private_download_' . uniqid() . '.txt';
            $this->client->fileDownloadSync($result->dataMap, $tempDownload);

            $this->assertFileExists($tempDownload);
            $this->assertEquals($content, file_get_contents($tempDownload));

            unlink($tempDownload);
        } finally {
            unlink($tempUpload);
        }
    }

    // =========================================================================
    // Error Handling Tests
    // =========================================================================

    public function testGetNonExistentPointer(): void
    {
        $ownerKey = SecretKey::random();
        $address = PointerAddress::fromOwner($ownerKey->publicKey());

        $this->expectException(\AntFfi\AntFfiException::class);
        $this->client->pointerGetSync($address);
    }

    public function testGetNonExistentScratchpad(): void
    {
        $ownerKey = SecretKey::random();
        $address = ScratchpadAddress::fromOwner($ownerKey->publicKey());

        $this->expectException(\AntFfi\AntFfiException::class);
        $this->client->scratchpadGetSync($address);
    }

    public function testGetNonExistentRegister(): void
    {
        $ownerKey = SecretKey::random();
        $address = RegisterAddress::fromOwner($ownerKey->publicKey());

        $this->expectException(\AntFfi\AntFfiException::class);
        $this->client->registerGetSync($address);
    }

    public function testGetNonExistentGraphEntry(): void
    {
        $ownerKey = SecretKey::random();
        $address = GraphEntryAddress::fromOwner($ownerKey->publicKey());

        $this->expectException(\AntFfi\AntFfiException::class);
        $this->client->graphEntryGetSync($address);
    }
}

File: dart/ant_ffi/test/client_integration_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:ant_ffi/ant_ffi.dart';

/// Default Anvil/Hardhat test private key #0 (pre-funded on local testnets)
const defaultTestKey =
    '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';

/// End-to-end integration tests for client operations.
///
/// These tests require a local Autonomi testnet running.
/// Run with: dart test --tags integration
///
/// To start a local testnet:
/// cd C:\tools\ant && setup-testnet.bat
@Tags(['integration'])
void main() {
  late Client client;
  late Wallet wallet;
  late Network network;

  setUpAll(() async {
    try {
      client = await Client.initLocal();
      network = Network.local();
      wallet = Wallet.fromPrivateKey(network, defaultTestKey);
    } catch (e) {
      fail('Local testnet not available: $e');
    }
  });

  tearDownAll(() {
    wallet.dispose();
    network.dispose();
    client.dispose();
  });

  group('Public Data Operations', () {
    test('dataPutPublic and dataGetPublic roundtrip', () async {
      final data = 'Hello, Autonomi Network! ${DateTime.now().millisecondsSinceEpoch}';

      final result = await client.dataPutPublic(data, wallet);

      expect(result.address.toHex(), isNotEmpty);
      expect(result.cost, isNotEmpty);

      // Retrieve the data
      final retrieved = await client.dataGetPublic(result.address.toHex());
      expect(retrieved, equals(data));

      result.address.dispose();
    });

    test('dataCost returns valid cost estimate', () async {
      final data = Uint8List.fromList(List.filled(1000, 0x58)); // 1KB of 'X'

      final cost = await client.dataCost(data);

      expect(cost, isNotEmpty);
    });
  });

  group('Private Data Operations', () {
    test('dataPut and dataGet roundtrip', () async {
      final data = Uint8List.fromList(
        utf8.encode('Secret data: ${DateTime.now().millisecondsSinceEpoch}'),
      );

      final result = await client.dataPut(data, wallet);

      expect(result.dataMap, isNotNull);
      expect(result.cost, isNotEmpty);

      // Retrieve the private data
      final retrieved = await client.dataGet(result.dataMap);
      expect(retrieved, equals(data));

      result.dataMap.dispose();
    });
  });

  group('Chunk Operations', () {
    test('chunkPut and chunkGet roundtrip', () async {
      final data = Uint8List.fromList(
        utf8.encode('Chunk content: ${DateTime.now().millisecondsSinceEpoch}'),
      );

      final result = await client.chunkPut(data, wallet);

      expect(result.address.toHex(), isNotEmpty);
      expect(result.cost, isNotEmpty);

      // Retrieve the chunk
      final chunk = await client.chunkGet(result.address);
      expect(chunk.value(), equals(data));

      chunk.dispose();
      result.address.dispose();
    });
  });

  group('Pointer Operations', () {
    test('pointerPut and pointerGet roundtrip', () async {
      final ownerKey = SecretKey.random();

      // Create a chunk address to point to
      final chunkAddress = ChunkAddress.fromContent(
        Uint8List.fromList(utf8.encode('target content')),
      );
      final target = PointerTarget.chunk(chunkAddress);

      // Create and store the pointer
      final pointer = NetworkPointer.create(ownerKey, 0, target);
      await client.pointerPut(pointer, wallet);

      // Retrieve the pointer
      final address = pointer.address();
      final retrieved = await client.pointerGet(address);

      expect(retrieved.counter(), equals(0));

      retrieved.dispose();
      address.dispose();
      pointer.dispose();
      target.dispose();
      chunkAddress.dispose();
      ownerKey.dispose();
    });

    test('pointer update with higher counter', () async {
      final ownerKey = SecretKey.random();

      // Create initial pointer
      final target1 = PointerTarget.chunk(
        ChunkAddress.fromContent(Uint8List.fromList(utf8.encode('target1'))),
      );
      final pointer1 = NetworkPointer.create(ownerKey, 0, target1);
      await client.pointerPut(pointer1, wallet);

      // Update the pointer with higher counter
      final target2 = PointerTarget.chunk(
        ChunkAddress.fromContent(Uint8List.fromList(utf8.encode('target2'))),
      );
      final pointer2 = NetworkPointer.create(ownerKey, 1, target2);
      await client.pointerPut(pointer2, wallet);

      // Retrieve and verify the updated pointer
      final address = pointer2.address();
      final retrieved = await client.pointerGet(address);

      expect(retrieved.counter(), equals(1));

      retrieved.dispose();
      address.dispose();
      pointer2.dispose();
      pointer1.dispose();
      target2.dispose();
      target1.dispose();
      ownerKey.dispose();
    });
  });

  group('Scratchpad Operations', () {
    test('scratchpadPut and scratchpadGet roundtrip', () async {
      final ownerKey = SecretKey.random();
      final data = Uint8List.fromList(
        utf8.encode('Scratchpad content: ${DateTime.now().millisecondsSinceEpoch}'),
      );

      // Create and store the scratchpad
      final scratchpad = Scratchpad.create(ownerKey, 1, data, 0);
      await client.scratchpadPut(scratchpad, wallet);

      // Retrieve the scratchpad
      final address = scratchpad.address();
      final retrieved = await client.scratchpadGet(address);

      expect(retrieved.counter(), equals(0));
      expect(retrieved.dataEncoding(), equals(1));

      // Decrypt and verify data
      final decrypted = retrieved.decryptData(ownerKey);
      expect(decrypted, equals(data));

      retrieved.dispose();
      address.dispose();
      scratchpad.dispose();
      ownerKey.dispose();
    });

    test('scratchpad update with higher counter', () async {
      final ownerKey = SecretKey.random();

      // Create initial scratchpad
      final scratchpad1 = Scratchpad.create(
        ownerKey,
        1,
        Uint8List.fromList(utf8.encode('initial data')),
        0,
      );
      await client.scratchpadPut(scratchpad1, wallet);

      // Update the scratchpad with higher counter
      final updatedData = Uint8List.fromList(utf8.encode('updated data'));
      final scratchpad2 = Scratchpad.create(ownerKey, 1, updatedData, 1);
      await client.scratchpadPut(scratchpad2, wallet);

      // Retrieve and verify the updated scratchpad
      final address = scratchpad2.address();
      final retrieved = await client.scratchpadGet(address);

      expect(retrieved.counter(), equals(1));
      expect(retrieved.decryptData(ownerKey), equals(updatedData));

      retrieved.dispose();
      address.dispose();
      scratchpad2.dispose();
      scratchpad1.dispose();
      ownerKey.dispose();
    });
  });

  group('Register Operations', () {
    test('registerCreate and registerGet roundtrip', () async {
      final ownerKey = SecretKey.random();
      final value = Uint8List.fromList(
        utf8.encode('Register value: ${DateTime.now().millisecondsSinceEpoch}'),
      );

      // Create the register
      final address = await client.registerCreate(value, ownerKey, wallet);

      expect(address.toHex(), isNotEmpty);

      // Retrieve the register value
      final retrieved = await client.registerGet(address);
      expect(retrieved, equals(value));

      address.dispose();
      ownerKey.dispose();
    });

    test('registerUpdate updates register value', () async {
      final ownerKey = SecretKey.random();
      final initialValue = Uint8List.fromList(utf8.encode('initial value'));

      // Create the register
      final address = await client.registerCreate(initialValue, ownerKey, wallet);

      // Update the register
      final updatedValue = Uint8List.fromList(utf8.encode('updated value'));
      await client.registerUpdate(updatedValue, ownerKey, wallet);

      // Retrieve and verify the updated value
      final retrieved = await client.registerGet(address);
      expect(retrieved, equals(updatedValue));

      address.dispose();
      ownerKey.dispose();
    });
  });

  group('Graph Entry Operations', () {
    test('graphEntryPut and graphEntryGet roundtrip', () async {
      final ownerKey = SecretKey.random();
      final content = Uint8List.fromList(
        utf8.encode('Graph entry content: ${DateTime.now().millisecondsSinceEpoch}'),
      );

      // Create and store the graph entry
      final entry = GraphEntry.create(
        ownerKey,
        Uint8List(0), // empty parents
        content,
        Uint8List(0), // empty descendants
      );
      await client.graphEntryPut(entry, wallet);

      // Retrieve the graph entry
      final address = entry.address();
      final retrieved = await client.graphEntryGet(address);

      expect(retrieved.content(), equals(content));

      retrieved.dispose();
      address.dispose();
      entry.dispose();
      ownerKey.dispose();
    });
  });

  group('Vault Operations', () {
    test('vaultPut and vaultGet user data roundtrip', () async {
      final vaultKey = VaultSecretKey.random();
      final userData = UserData.create();

      // Store user data to the vault
      await client.vaultPut(vaultKey, userData, wallet);

      // Retrieve user data from the vault
      final retrieved = await client.vaultGet(vaultKey);

      expect(retrieved, isNotNull);

      retrieved.dispose();
      userData.dispose();
      vaultKey.dispose();
    });
  });

  group('Archive Operations', () {
    test('archivePutPublic and archiveGetPublic roundtrip', () async {
      // Create an archive with a file
      var archive = PublicArchive.create();

      // Upload some data first to get an address
      final result = await client.dataPutPublic('file content', wallet);

      // Add the file to the archive
      final metadata = Metadata.create(12);
      archive = archive.addFile('/test/file.txt', result.address, metadata);

      expect(archive.fileCount(), equals(1));

      // Store the archive
      final address = await client.archivePutPublic(archive, wallet);

      expect(address.toHex(), isNotEmpty);

      // Retrieve the archive
      final retrieved = await client.archiveGetPublic(address);

      expect(retrieved.fileCount(), equals(1));

      retrieved.dispose();
      address.dispose();
      archive.dispose();
      metadata.dispose();
      result.address.dispose();
    });
  });

  group('Error Handling', () {
    test('get non-existent pointer throws', () async {
      final ownerKey = SecretKey.random();
      final address = PointerAddress.fromPublicKey(ownerKey.publicKey());

      expect(
        () async => await client.pointerGet(address),
        throwsA(isA<AntFfiException>()),
      );

      address.dispose();
      ownerKey.dispose();
    });

    test('get non-existent scratchpad throws', () async {
      final ownerKey = SecretKey.random();
      final address = ScratchpadAddress.fromPublicKey(ownerKey.publicKey());

      expect(
        () async => await client.scratchpadGet(address),
        throwsA(isA<AntFfiException>()),
      );

      address.dispose();
      ownerKey.dispose();
    });

    test('get non-existent register throws', () async {
      final ownerKey = SecretKey.random();
      final address = RegisterAddress.fromOwner(ownerKey.publicKey());

      expect(
        () async => await client.registerGet(address),
        throwsA(isA<AntFfiException>()),
      );

      address.dispose();
      ownerKey.dispose();
    });

    test('get non-existent graph entry throws', () async {
      final ownerKey = SecretKey.random();
      final address = GraphEntryAddress.fromOwner(ownerKey.publicKey());

      expect(
        () async => await client.graphEntryGet(address),
        throwsA(isA<AntFfiException>()),
      );

      address.dispose();
      ownerKey.dispose();
    });
  });
}
